transport-native-kqueue/src/main/java/io/netty/channel/kqueue/AbstractKQueueChannel.java

```diff
@@ -428,6 +424,23 @@ final void readReadyFinally(ChannelConfig config) {
             }
         }
 
+        final boolean failConnectPromise(Throwable cause) {
+            if (connectPromise != null) {
+                // SO_ERROR has been shown to return 0 on macOS if detect an error via read() and the write filter was
+                // not set before calling connect. This means finishConnect will not detect any error and would
+                // successfully complete the connectPromise and update the channel state to active (which is incorrect).
+                ChannelPromise connectPromise = AbstractKQueueChannel.this.connectPromise;
+                AbstractKQueueChannel.this.connectPromise = null;
+                ConnectException e = new ConnectException("failed to connect");
```

Comment:
Should we check first if cause is already a ConnectException ?

Reply:
done (in practice it isn't though).